machineconfiguration/v1/types.go

```diff
@@ -0,0 +1,917 @@
+
+	// message provides additional information about the current condition.
+	// This is only to be consumed by humans.
+	// +optional
+	Message string `json:"message,omitempty"`
+}
+
+// ContainerRuntimeConfigStatusConditionType is the state of the operator's reconciliation functionality.
+type ContainerRuntimeConfigStatusConditionType string
+
+const (
+	// ContainerRuntimeConfigSuccess designates a successful application of a ContainerRuntimeConfig CR.
+	ContainerRuntimeConfigSuccess ContainerRuntimeConfigStatusConditionType = "Success"
+
+	// ContainerRuntimeConfigFailure designates a failure applying a ContainerRuntimeConfig CR.
+	ContainerRuntimeConfigFailure ContainerRuntimeConfigStatusConditionType = "Failure"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ContainerRuntimeConfigList is a list of ContainerRuntimeConfig resources
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type ContainerRuntimeConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []ContainerRuntimeConfig `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineState describes the health of the Machines on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineState struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec MachineStateSpec `json:"spec"`
+	// +optional
+	Status MachineStateStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineStateList describes all of the MachinesStates on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineStateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineState `json:"items"`
+}
+
+// MachineStateSpec describes the type of State we are managing
+type MachineStateSpec struct {
+	// Kind describes the type of MachineState
+	// +optional
+	Kind   MachineStateType   `json:"machineStateKind,omitempty"`
+	Config MachineStateConfig `json:"configuration"`
+}
+
+// MachineStateStatus holds the reported information on a particular MachineState
+type MachineStateStatus struct {
+	// Config Describes the Kind of MachineState
+	Config MachineStateConfig `json:"configuration"`
+	// MostRecentState MCO Object Mapped to a Progress Condition
+	MostRecentState map[string]ProgressionCondition `json:"progress"`
+	// Progression describes the 20 Most Recent MachineStates
```

Comment:
nit: Do we want to commit to the 20 most recent?  Or can we be handwave with an unqualified "recent"?  E.g. see openshift/cluster-version-operator#805 where we landed some fancy ClusterVersion `history` pruning to try to balance `status` size vs. the more interesting bits of the previous history.  I could see a MachineConfigPool with a lot of nodes history like:

* 99 events ago, we finished updating this pool to `rendered-*`.
* 98 events ago, we started updating this pool to `rendered-*`.
* 97 events ago, we cordoned node A.
* 96 events ago, we drained node A.
* ...
* here we are, stuck draining node XYZ.

where the most-interesting events would be the pool-level transitions and a few of the more recent node events, but not anything about node A which was quite happy a long time ago.  I realize that that kind of pruning takes effort to develop, and "most recent 20" is a good first-pass at a pruning algorithm.   And maybe it will be sufficient forever.  But I don't see a problem with relaxing the docstring here a bit to leave the door open, in case we do decide we want to adjust the pruning algorithm later.

nit: I don't understand the Title Casing here.  Can we explain the semantics without needing to refer to concepts by title?

nit: It's a bit hard to find this in the pull-request diff.  Thoughts about sharding into per-CRD-level-type files like [`config/v1`'s `types_*.go`][1]?

[1]: https://github.com/openshift/api/tree/4c26c14d9ebbef4e5957da796bd7b8fb0401c71c/config/v1

Reply:
I think I am opting to change this map to omit the amount it tracks and leave that upto the MCO who populates the field.